nifi-nar-bundles/nifi-enrich-bundle/nifi-enrich-processors/src/main/java/org/apache/nifi/processors/GeoEnrichIPRecord.java

```diff
@@ -0,0 +1,323 @@
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile input = session.get();
+        if (input == null) {
+            return;
+        }
+
+        FlowFile output = session.create(input);
+        FlowFile notFound = splitOutput ? session.create(input) : null;
+        final DatabaseReader dbReader = databaseReaderRef.get();
+        try (InputStream is = session.read(input);
+             OutputStream os = session.write(output);
+             OutputStream osNotFound = splitOutput ? session.write(notFound) : null) {
+            RecordPathCache cache = new RecordPathCache(GEO_PROPERTIES.size() + 1);
+            Map<PropertyDescriptor, RecordPath> paths = new HashMap<>();
+            for (PropertyDescriptor descriptor : GEO_PROPERTIES) {
+                if (!context.getProperty(descriptor).isSet()) {
+                    continue;
+                }
+                String rawPath = context.getProperty(descriptor).evaluateAttributeExpressions(input).getValue();
+                RecordPath compiled = cache.getCompiled(rawPath);
+                paths.put(descriptor, compiled);
+            }
+
+            String rawIpPath = context.getProperty(IP_RECORD_PATH).evaluateAttributeExpressions(input).getValue();
+            RecordPath ipPath = cache.getCompiled(rawIpPath);
+
+            RecordReader reader = readerFactory.createRecordReader(input, is, getLogger());
+            RecordSchema schema = writerFactory.getSchema(input.getAttributes(), null);
+            RecordSetWriter writer = writerFactory.createWriter(getLogger(), schema, os);
+            RecordSetWriter notFoundWriter = splitOutput ? writerFactory.createWriter(getLogger(), schema, osNotFound) : null;
+            Record record;
+            Relationship targetRelationship = REL_NOT_FOUND;
+            writer.beginRecordSet();
+
+            while ((record = reader.nextRecord()) != null) {
+                CityResponse response = geocode(ipPath, record, dbReader);
+                boolean wasEnriched = enrichRecord(response, record, paths);
+                if (wasEnriched) {
+                    targetRelationship = REL_FOUND;
+                }
+                if (!splitOutput || (splitOutput && wasEnriched)) {
+                    writer.write(record);
+                } else {
+                    notFoundWriter.write(record);
+                }
+            }
+            writer.finishRecordSet();
+            writer.close();
+            is.close();
+            os.close();
+            if (osNotFound != null) {
+                osNotFound.close();
+            }
+
+            if (!splitOutput) {
+                session.transfer(output, targetRelationship);
+                session.remove(input);
+            } else {
+                session.transfer(output, REL_FOUND);
+                session.transfer(notFound, REL_NOT_FOUND);
```

Comment:
the `notFound` flow file might have 0 records in it, should we still send out an empty flow file if so? Right now it does, but I think it should either not send an empty flow file out, or have it be configurable